src/service/viewport-impl.js

```diff
@@ -784,6 +793,144 @@ export class ViewportBindingNaturalIosEmbed_ {
+      for (let i = 0; i < elements.length; i++) {
+        if (i > 10) {
+          // DO NOT SUBMIT: some constraint may be needed.
+          break;
+        }
+        const element = elements[i];
+        if (element.hasAttribute('i-amp-fixedid')) {
+          // Already seen.
+          continue;
+        }
+        const fixedId = 'F' + this.fixedElements_.length;
+        element.setAttribute('i-amp-fixedid', fixedId);
+        this.fixedElements_.push({id: fixedId, element: element});
+      }
+    });
+    this.updateFixedElements_();
+  }
+
+  /**
+   * @param {!Array<CSSRule>} rules
+   * @param {!Array<string>} foundSelectors
+   * @private
+   */
+  discoverFixedSelectors_(rules, foundSelectors) {
+    for (let i = 0; i < rules.length; i++) {
+      const rule = rules[i];
+      if (rule.type == /* CSSStyleRule */ 1) {
+        if (rule.selectorText != '*' && rule.style.position == 'fixed') {
+          foundSelectors.push(rule.selectorText);
+        }
+      } else if (rule.type == /* CSSMediaRule */ 4) {
+        this.discoverFixedSelectors_(rule.cssRules, foundSelectors);
+      } else if (rule.type == /* CSSSupportsRule */ 12) {
+        this.discoverFixedSelectors_(rule.cssRules, foundSelectors);
+      }
+    }
+  }
+
+  /**
+   * Moves elements between fixed layer and BODY depending on whether they are
+   * currently visible.
+   * @return {!Promise}
+   * @private
+   */
+  updateFixedElements_() {
+    if (this.fixedElements_.length == 0) {
+      return Promise.resolve();
+    }
+    // DO NOT SUBMIT: remove all "removed" elements.
+    return this.vsync_.runPromise({
+      measure: state => {
+        state.fixed = [];
+        this.fixedElements_.forEach(fe => {
+          const styles = this.win.getComputedStyle(fe.element, null);
+          // DO NOT SUBMIT: calculate zIndex based on the existing zIndex and
+          // document position.
+          state.fixed.push(styles.getPropertyValue('position') == 'fixed');
+        });
+      },
+      mutate: state => {
+        this.fixedElements_.forEach((fe, i) => {
+          const element = fe.element;
+          const oldFixed = element.parentElement == this.fixedLayer_;
+          const newFixed = state.fixed[i];
+          if (oldFixed != newFixed) {
+            if (newFixed) {
+              // Move element to the fixed layer.
+              if (!fe.placeholder) {
+                // Never been added before: ensure that it's properly configured.
+                setStyles(element, {
+                  pointerEvents: 'initial'
+                });
+                fe.placeholder = this.win.document.createElement('i-amp-fp');
+                fe.placeholder.setAttribute('i-amp-fixedid', fe.id);
+                setStyles(fe.placeholder, {
+                  display: 'none'
+                });
+              }
+              element.parentElement.replaceChild(fe.placeholder, element);
+              this.fixedLayer_.appendChild(element);
```

Comment:
Hmm. Wasn't planning on this, but I probably should. WDYT? It seems the only way is to check and push it back in DOM if it breaks.